### DISCUSSION TITLE
backpack/calendar-operation - Performance Improvement in Month View Rendering

### DISCUSSION THREAD

**Discussion by tomsb** [Laravel-Backpack/community-forum#1265] [General]
I have approximately 150 events in the month view. Previously, it took about **3-4 seconds** to render them.  

With these changes, the render time has been reduced to around **1 second**, representing a **3-4x performance improvement**. This difference is highly noticeable.

#### File: `calendar_view.blade.php` (From line 218)
```diff
event?.remove();
- calendar.addEvent(eventData);
- applySearch();
// ....
+ .then(data => {
+     if (searchInput.value === ''){
+          calendar.addEventSource(events);
+      }
+      else{
+          applySearch();
+      }
+  })
```
Calendar events are now added in one batch
Search is only executed if there is something to search for, whereas previously it was called unnecessarily every time.

What do you think?

[a comment by an automated account is not shown]

**Comment by karandatwani92** [comment]
Hey @tomsb 

Thanks for the excellent suggestion.👍
@pxpm What do you think on this?

**Comment by pxpm** [comment]
Thanks @tomsb I think it's a great addition, thanks for the suggestion 👍 I've just released a new calendar version with that improvement. Please allow a few moments for our private repository to pick the changes and you can get the updated version with a `composer update`. 

Cheers 🙏

**Comment by tomsb** [comment]
This optimization has introduced a small bug.

### Steps to Reproduce
1. Ensure the calendar view is "larger than the screen" so that you need to scroll the browser scrollbar down to see the bottom of the calendar.
2. Open another browser tab.
3. Return to the calendar tab.

### Observed Behavior
The browser scrollbar is scrolled back to the top.

### Expected Behavior
The scrollbar should stay in its previous position and not scroll to the top.

### Root Cause
This happens because a `visibilitychange` event is triggered in the background when switching tabs. This event reloads the calendar data and causes the calendar to redraw, which results in the scrollbar resetting to the top.

### Attempts to Fix
I have tried several approaches, but none have fully resolved the issue. I did find one solution that works, but it is not elegant:

```diff
                // ...
		event?.remove();
+		if (isTriggeredByPageVisibilityChange) {
+			calendar.addEvent(eventData);
+			if (searchInput.value && searchInput.value !== '') {
+				applySearch();
+			}
+		}
	});

+	if (!isTriggeredByPageVisibilityChange) {
		calendar.addEventSource(events);
		if (searchInput.value && searchInput.value !== '') {
			applySearch();
		}
+	}

+	isTriggeredByPageVisibilityChange = false;
});

// ...

+ let isTriggeredByPageVisibilityChange = false;
  function pageVisibilitychange() {
    if (document.visibilityState === 'visible') {
        hideMenus();
+       isTriggeredByPageVisibilityChange = true;
        datesSet();
    }
  }
```
